src/vmspawn/vmspawn-util.c

```diff
@@ -0,0 +1,195 @@
+
+                        if (ret_firmware_path)
+                                *ret_firmware_path = firmware;
+                        return 1;
+                }
+        }
+
+#undef FIRMWARE_LOCATIONS
+        #define FIRMWARE_LOCATIONS \
+                "/usr/share/edk2/ovmf/OVMF_CODE.secboot.fd", \
+                "/usr/share/edk2-ovmf/OVMF_CODE.secboot.fd", \
+                "/usr/share/qemu/OVMF_CODE.secboot.fd", \
+                "/usr/share/ovmf/OVMF.secboot.fd", \
+                "/usr/share/OVMF/OVMF_CODE.secboot.fd"
+
+        FOREACH_STRING(firmware, FIRMWARE_LOCATIONS) {
+                _cleanup_close_ int fd = -EBADF;
+                fd = open(firmware, O_CLOEXEC | O_RDONLY);
+                if (fd >= 0) {
+                        if (ret_firmware_path)
+                                *ret_firmware_path = firmware;
+                        return 0;
+                }
+        }
+
+#undef FIRMWARE_LOCATIONS
+        #define FIRMWARE_LOCATIONS \
+                "/usr/share/edk2/ovmf/OVMF_CODE.fd", \
+                "/usr/share/edk2-ovmf/OVMF_CODE.fd", \
+                "/usr/share/qemu/OVMF_CODE.fd", \
+                "/usr/share/ovmf/OVMF.fd", \
+                "/usr/share/OVMF/OVMF_CODE.fd"
+
+        FOREACH_STRING(firmware, FIRMWARE_LOCATIONS) {
+                _cleanup_close_ int fd = -EBADF;
+                fd = open(firmware, O_CLOEXEC | O_RDONLY);
+                if (fd >= 0) {
+                        log_warning("Couldn't find OVMF firmware blob with secure boot support, "
+                                    "falling back to OVMF firmware blobs without secure boot support.");
+
+                        if (ret_firmware_path)
+                                *ret_firmware_path = firmware;
+                        return 1;
+                }
+        }
+
+#undef FIRMWARE_LOCATIONS
+        if (ret_firmware_path)
+                *ret_firmware_path = NULL;
+        log_error("Couldn't find OVMF UEFI firmware blob.");
+        return -ENOENT;
+}
+
+int find_qemu_binary(char **ret_qemu_binary) {
+        int r;
+
+        static const char *architecture_to_qemu_table[_ARCHITECTURE_MAX] = {
+                [ARCHITECTURE_ARM64]       = "aarch64",
+                [ARCHITECTURE_ARM]         = "arm",
+                [ARCHITECTURE_ALPHA]       = "alpha",
+                [ARCHITECTURE_X86_64]      = "x86_64",
+                [ARCHITECTURE_X86]         = "i386",
+                [ARCHITECTURE_LOONGARCH64] = "loongarch64",
+                [ARCHITECTURE_MIPS64_LE]   = "mips",
+                [ARCHITECTURE_MIPS_LE]     = "mips",
+                [ARCHITECTURE_PARISC]      = "hppa",
+                [ARCHITECTURE_PPC64_LE]    = "ppc",
+                [ARCHITECTURE_PPC64]       = "ppc",
+                [ARCHITECTURE_PPC]         = "ppc",
+                [ARCHITECTURE_RISCV32]     = "riscv32",
+                [ARCHITECTURE_RISCV64]     = "riscv64",
+                [ARCHITECTURE_S390X]       = "s390x",
```

Comment:
might be nice adding some comment such as `/* differs from our names! */` or so for the cases where the names qemu uses differ from our names.